scripts/traccc_throughput_mt_profiler.sh

```diff
@@ -0,0 +1,150 @@
+         ;;
+      h)
+         usage
+         exit 0
+         ;;
+      :)
+         echo "Argument -$OPTARG requires a parameter!"
+         usage
+         exit 1
+         ;;
+      ?)
+         echo "Unknown argument: -$OPTARG"
+         usage
+         exit 1
+         ;;
+   esac
+done
+
+# Print the configuration received.
+echo "Using configuration:"
+echo "   EXECUTABLE  : ${TRACCC_EXECUTABLE}"
+echo "   MAX_THREADS : ${TRACCC_MAX_THREADS}"
+echo "   REPETITIONS : ${TRACCC_REPETITIONS}"
+echo "   EVT_MULTI   : ${TRACCC_EVT_MULTI}"
+echo "   CSV_FILE    : ${TRACCC_CSV_FILE}"
+echo "   DET_FILE    : ${TRACCC_DET_FILE}"
+echo "   DIGI_FILE   : ${TRACCC_DIGI_FILE}"
+
+# Check whether the output file already exists. Refuse to overwrite existing
+# files.
+if [[ -f "${TRACCC_CSV_FILE}" ]]; then
+   echo "***"
+   echo "*** Will not overwrite ${TRACCC_CSV_FILE}!"
+   echo "***"
+   exit 1
+fi
+
+# The input directories to use.
+TRACCC_INPUT_DIRS=("ttbar_mu20"  "ttbar_mu40"  "ttbar_mu60"  "ttbar_mu80"
+                   "ttbar_mu100" "ttbar_mu140" "ttbar_mu200" "ttbar_mu300")
+
+# The number of events to process for the different mu values. Chosen to take
+# roughly the same amount of time to process on a CPU.
+declare -A TRACCC_EVT_COUNT
+TRACCC_EVT_COUNT["ttbar_mu20"]=$((1000*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu40"]=$((500*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu60"]=$((250*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu80"]=$((200*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu100"]=$((150*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu140"]=$((90*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu200"]=$((50*${TRACCC_EVT_MULTI}))
+TRACCC_EVT_COUNT["ttbar_mu300"]=$((20*${TRACCC_EVT_MULTI}))
+
+# Put a header on the CSV file.
+echo "directory,threads,loaded_events,cold_run_events,processed_events,warm_up_time,processing_time" \
+   > "${TRACCC_CSV_FILE}"
+
+# Counter for a nice printout.
+COUNTER=1
+COUNT=$((${#TRACCC_INPUT_DIRS[@]}*${TRACCC_MAX_THREADS}*${TRACCC_REPETITIONS}))
+
+# Iterate over the number of threads.
+for NTHREAD in $(seq 1 ${TRACCC_MAX_THREADS}); do
+   # Iterate over the input datasets.
+   for EVTDIR in ${TRACCC_INPUT_DIRS[@]}; do
+      # Perform the requested number of repetitions.
+      for REPEAT in $(seq 1 ${TRACCC_REPETITIONS}); do
+
+         # Tell the user what's happening.
+         echo ""
+         echo "Running test ${COUNTER} / ${COUNT}"
+         ((COUNTER++))
+
+         # Run the throughput test.
+         ${TRACCC_EXECUTABLE}                                                 \
+            --detector_file="${TRACCC_DET_FILE}"                              \
+            --digitization_config_file="${TRACCC_DIGI_FILE}"                  \
+            --input_directory="tml_full/${EVTDIR}/"                           \
+            --threads=${NTHREAD}                                              \
+            --cold_run_events=$((20*${NTHREAD}))                              \
```

Comment:
Why? :confused: The purpose of the warm-up events is to:
  - Make sure that the kernels are loaded into the device;
  - Warm up the memory caches.

Why would either of these depend on the pileup value? I chose 20 because this gives a good chance of encountering all 10 of our events at least once on each thread. One could increase this a bit further, but this seemed to be "good enough".

Reply:
I would identify two reasons; first, smaller events will fill up less of the caches, so if you want to get comparable caching effects you will want to preprocess more of them. More importantly, the warm-up events serve to allow the device to scale up its power states and stabilize its execution frequency. A small number of small events will do a worse job at this than a small number of large events, simply due to the difference in execution time.

As far as I am concerned, the way @guilhermeAlmeida1 does this is the way to go. I think he takes 10% of the total number of events and runs them as warm-up events. That way, as long as the total event count is scaled properly to the pile-up, the warm-up will too.

Reply:
That is indeed how I do things. Your logic sounds correct but I admit my logic mostly came just from empirical data where I generally get more "stable" results this way.

Reply:
Yepp, and you guys should implement all of this in a next step. :+1: For the <% effects that this includes (For small events we run many of them anyway. The warm-up was resulting in pretty negligible effects in my tests for those events.) I was not willing to spend more time on this.